Optimization of camera mode change for better E-M10 MKIV support

embedded-ideas opened the pull request:
The current code does some unmanaged switching between different camera modes. This has some drawbacks especially for the E-M10 MKIV which doesn't have a "SHUTTER" mode. In order to take a picture with the E-M10 MKIV you have to stay in "RECORD" mode with live view enabled. Using the code provided by camdroid (https://github.com/joergmlpts/olympus-wifi/pull/3) it is possible to take pictures, but there are two restarts of the liveview involved. Further you have to know that you need an instance of the M10MKIV class.

What I tried to archive with the code changes is to add an approach that allows the execution of one action at a time with a mode change only if it is required. For example: If you want to set the time while you are in the PLAY mode, the mode will not be changed. If you currently execute the live view in RECORD mode, the mode will be temporarily changed during the operation. Live view will be restarted automatically. 

Besides a faster operation, this allows the E-M10 MKIV to be efficiently used with the rest of the code. I added an extra path for taking pictures with the E-10 MKIV instead of the inherited class. If there are more differences with other new camera models it might be a good idea for the previous approach, but I think this should be used in combination with a factory of some kind. At this stage I think it doesn't need that kind of complexity.

I hope the code quality is sufficient as I am normally not using Python on a daily basis ;)

I tested the code using my E-M10 MKIV. I don't own any other model. It would be good if anyone is able to do some testing using an older model which supports the SHUTTER mode.

joergmlpts commented:
Thanks for the changes! I tried them on my TG-5.

I started the live view. The Live View window started big, in the highest resolution. I took a photo. Taking the photo worked fine, but after the photo was taken, the live view window became small. It was then displayed in the lowest resolution. The desired size in the menu remained at the highest resolution.

Do you know why this happens? I undid the changes and verified that without your changes, the window size does not change when a photo is taken.

embedded-ideas commented:
Thank you very much for the testing. I tried to reproduce the behavior on my side. Unfortunately I wasn't able simulate the problem with the E-M10. I tried different resolutions and switched to a different mode. The code always switched back to the right resolution.

I tested with a mode switch to the "play" mode by setting the clock.

![liveview_res_problem](https://github.com/user-attachments/assets/f36d914c-8f07-474f-8cc0-e86626c94b6e)

1. Setting the time after liveview has been opened ("File->Set clock") -> causes mode change to "play"
2. Switch to a different resolution via "View->Size->640x480"
3. Setting the time again

The network log is coherent with what I see on the screen. I also tested if a switch to a different mode implicitly stops the liveview (it does) and if the liveview is able to restart (it is). 

During testing I accidentally added a mode change to "rec" right before starting the liveview, and bingo, it results in the same behavior as you described. 

![liveview_res_problem_lowres](https://github.com/user-attachments/assets/cc4f940d-f334-42b9-8bd0-41e26d74c00b)

I would interpret this as follows: A mode change request to the same mode will be ignored by the camera. Switching to "rec" without a parameter will set the resolution to the lowest possible. This causes the liveview to be started in a low res.

Maybe the shutter mode and the rec mode are somehow "related". I'm not able to test it with the E-M10 so I added a fix which will ensure that starting liveview will always begin from "play" mode. This will fix at least the observed problem when switching from "rec" to "rec" mode.

joergmlpts commented:
Thanks! The window size now stays the same when I take a photo.
